### PR TITLE
Save/load entire model & model config

### PR DESCRIPTION
**Pull request by warshallrho**
### Checklist
- [x] I've tested that my changes are compatible with the latest version of Tensorflow.
- [x] I've read the [Contribution Guidelines](https://github.com/tensorlayer/tensorlayer/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

Update get_start_advance.rst

### Description

Update get_start_advance.rst

### CODE REVIEW

File: docs/user/get_start_advance.rst [hunk 1]
```diff
@@ -13,20 +13,49 @@ Get entire CNN
...
+  import tensorlayer as tl
+  import numpy as np
+  from tensorlayer.models.imagenet_classes import class_names
+
+  vgg = tl.models.vgg16(pretrained=True)
+  img = tl.prepro.imresize(tl.vis.read_image('img.jpeg'), (224, 224)).astype(np.float32)
+  mean = np.array([123.68, 116.779, 103.939], dtype=np.float32).reshape([1, 1, 1, 3])
+  img = img - mean
+  vgg.eval()
+  output = vgg(img1)
```
Comment (JingqingZ) [inline docs/user/get_start_advance.rst]: Should be `output=vgg(img)`

File: docs/user/get_start_advance.rst [hunk 2]
```diff
@@ -13,20 +13,49 @@ Get entire CNN
...
 .. code-block:: python
 
-  xxx
+  import tensorflow as tf
+  import tensorlayer as tl
 
+  # get VGG without the last layer
+  vgg = tl.models.vgg.vgg16(end_with='fc2_relu')
+  # add one more layer
+  net = tl.layers.Dense(n_units=100, name='out')(vgg)
```
Comment (JingqingZ) [inline docs/user/get_start_advance.rst]: A Model cannot be fed into a Layer.

File: docs/user/get_start_advance.rst [hunk 3]
```diff
@@ -13,20 +13,49 @@ Get entire CNN
...
+  # restore pre-trained VGG parameters
+  vgg.restore_weights()
+  # train your own classifier (only update the last layer)
+  train_params = tl.layers.get_variables_with_name('out')
```
Comment (JingqingZ) [inline docs/user/get_start_advance.rst]: `tl.layers.get_variables_with_name()` has a `deprecated` label. It seems it has not been refactored (?)

Comment (ChrisWu1997) [reply docs/user/get_start_advance.rst]: It is deprecated and has been removed. To get a particular weight, now user need to use model.weights or model.all_layers to find it. Shall we implement a this `get_variables_with_name` as a method of Model?

**Comment by zsdonghao** [comment]
@warshallrho Should we change the input value range from 0~255 to 1~1 ?

File: docs/user/get_start_advance.rst (hunk 3 above)
Comment (JingqingZ) [reply docs/user/get_start_advance.rst]: Agree, a similar function is needed for Model.

Comment (ChrisWu1997) [reply docs/user/get_start_advance.rst]: I plan to implement `get_layer`, `get_weight` for Model. In that case, to only train fc layer of VGG, user can simply fetch fc weights by `train_params=vgg.get_layer('out').weights`.

**Comment by warshallrho** [comment]
> @warshallrho Please make changes after I finished `get_layer` and `get_weight`.

got it

File: tensorlayer/files/utils.py [hunk 4]
```diff
@@ -93,99 +117,200 @@ def save_graph(network=None, name='graph.pkl'):
...
-                func_name = fn_dict['func_name']
-                lib = importlib.import_module(module_path)
-                fn = getattr(lib, func_name)
-                layer_kwargs[key] = fn
-                # print(key, layer_kwargs[key])
-        # print(name, prev_layer, layer_class, layer_kwargs)
+    Examples
+    --------
+    Save the architecture
+    >>> tl.files.save_graph(net_test, 'graph.pkl')
```
Comment (ChrisWu1997) [inline tensorlayer/files/utils.py]: Here should be 'graph.h5'

File: tensorlayer/layers/core.py [hunk 5]
```diff
@@ -92,6 +95,17 @@ def __init__(self, name=None, *args, **kwargs):
...
+        self.graph.update({'class': self.__class__.__name__.split('.')[-1]})
+        self.layer_args = self._get_init_args(skip=3)
+        self.layer_args.update(self.get_args())
+        self.layer_args["name"] = self.name
+        self.graph.update({"args": self.layer_args})
+        if self.__class__.__name__ in tl.layers.inputs.__all__:
+            self.graph.update({'prev_layer': None})
+        else:
+            self.graph.update({'prev_layer': []})
+
```
Comment (ChrisWu1997) [inline tensorlayer/layers/core.py]: Is it possible that we move `graph` part out of `__init__`, e.g. make it a class method which should be only called in Model._construct_graph. Because for dynamic model, this `graph` part seems to be an extra burden for layer initialization, given that we do not support dynamic model save now.

File: tensorlayer/files/utils.py (hunk 4 above)
Comment (warshallrho) [reply tensorlayer/files/utils.py]: I haven't updated the doc yet, and we need to use model.save(filepath="h5.hdf5", save_weights=False) or tl.files.save_hdf5_graph(model, name='h5.hdf5', save_weights=False)

Comment (ChrisWu1997) [reply tensorlayer/files/utils.py]: I suggest no default value for `filepath` and `save_weights=True` after you complete all these. That makes more sense.

File: tensorlayer/layers/core.py (hunk 5 above)
Comment (warshallrho) [reply tensorlayer/layers/core.py]: I will try it tomorrow, but self.layer_args = self._get_init_args(skip=3) is needed in __init__(), as we can only get init_args in __init__()

File: tensorlayer/files/utils.py (hunk 4 above)
Comment (warshallrho) [reply tensorlayer/files/utils.py]: got it

File: tensorlayer/files/utils.py [hunk 6]
```diff
@@ -195,25 +278,47 @@ def load_graph(name='graph.pkl'):
...
-    outputs = _graph2net(saved_file["outputs"])
-    model_name = saved_file["name"]
-    from tensorlayer.models import Model
-    M = Model(inputs=inputs, outputs=outputs, name=model_name)
+    M = static_graph2net(saved_file)
+
     return M
 
 
+def save_pkl_graph(network, name='model.pkl'):
```
Comment (JingqingZ) [inline tensorlayer/files/utils.py]: `save_pkl_graph` is not ready yet? it seems it is never called.

File: tensorlayer/models/core.py [hunk 7]
```diff
@@ -661,20 +689,50 @@ def release_memory(self):
...
-    #         raise AssertionError(
-    #             "save_graph not support dynamic mode yet"
-    #         )
-    #     utils.save_graph(network=self, name=filepath)
-    #
-    #
-    # def load(filepath):
-    #     return utils.load_graph(name=filepath)
+    def save(self, filepath, save_weights=True):
+        """Save model into a given file. Support saving model weights.
```
Comment (JingqingZ) [inline tensorlayer/models/core.py]: The comment here needs to provide more information to emphasise the difference between `save` and `save_weights`. This function `save` can save both the architecture of neural networks and weights (optional).

File: tensorlayer/models/core.py [hunk 8]
```diff
@@ -661,20 +689,50 @@ def release_memory(self):
...
+        >>> new_net = Model.load('./model.h5', load_weights=True)
+        """
+        if self.outputs is None:
+            raise RuntimeError(
+                "Model save() not support dynamic mode yet.\nHint: you can use Model save_weights() to save the weights in dynamic mode."
+            )
+        utils.save_hdf5_graph(network=self, filepath=filepath, save_weights=save_weights)
+
+    @staticmethod
+    def load(filepath, load_weights=True):
```
Comment (JingqingZ) [inline tensorlayer/models/core.py]: The comment here needs to emphasise the difference between `load` and `load_weights`. This function `load` can load both the architecture of neural networks and weights.

File: tests/models/test_model_save_graph.py [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+
+class Lambda_layer_test(CustomTestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        print("##### begin testing lambda layer #####")
+
+    def test_list_inputs_outputs(self):
+        x = tl.layers.Input([8, 3], name='input')
+        y = tl.layers.Lambda(lambda x: 2*x, name='lambda')(x)
```
Comment (JingqingZ) [inline tests/models/test_model_save_graph.py]: Just want to confirm the progress of testing Lambda Layer:
1. non-parametric and non-args like `lambda x: 2*x`: tested
2. non-parametric and with args like `some_func(x, foo=1)`: not tested
3. parametric: e.g. lambda layer includes a keras model inside with trainable weights: not tested

File: tensorlayer/models/core.py (hunk 8 above)
Comment (JingqingZ) [reply tensorlayer/models/core.py]: And also emphasise when a model is loaded, there is no need to reimplement or declare the architecture of the model explicitly in code.

File: tests/models/test_model_save_graph.py (hunk 9 above)
Comment (warshallrho) [reply tests/models/test_model_save_graph.py]: I have got the above mentioned comments.
Rundi and I are trying to support lambda layer includes a keras model inside with trainable weights now.

File: tensorlayer/files/utils.py (hunk 6 above)
Comment (warshallrho) [reply tensorlayer/files/utils.py]: We do not need to use save_pkl_graph now, but we may use this function in the future, so I retain this function temporarily. It is not visible outside utils.py, as it is not in __all__.

File: tests/models/test_model_save_graph.py (hunk 9 above)
Comment (ChrisWu1997) [reply tests/models/test_model_save_graph.py]: > Just want to confirm the progress of testing Lambda Layer:
> 
> 1. non-parametric and non-args like `lambda x: 2*x`: tested
> 2. non-parametric and with args like `some_func(x, foo=1)`: not tested
> 3. parametric: e.g. lambda layer includes a keras model inside with trainable weights: not tested

@JingqingZ @zsdonghao , to summarize current functionality of model save/load Lambda Layer:
1. non-parametric and non-args like `lambda x: 2*x`: OK and tested.
2. non-parametric and with args like `some_func(x, foo=1)`: OK and not tested.
3. Keras model/layer with trainable weights: OK and tested.
4. parametric self defined function with outside variables: __not supported now__.
For example: 
```python
a = 5
def func(x):
     return x + a
```
Current version could not save outside variable `a`. As I think this is not a very common case, and dealing with this issue may requires much time, I think maybe we can skip this case for our alpha version release?

Comment (JingqingZ) [reply tests/models/test_model_save_graph.py]: > 2\. non-parametric and with args like `some_func(x, foo=1)`: OK and not tested.

Emm, is it a typo here? not tested or tested?

I agree, you may ignore the 4th case right now. Please write a 'TODO' in the code and a note in the documentation that warns users to avoid this case.

Comment (ChrisWu1997) [reply tests/models/test_model_save_graph.py]: > > 2. non-parametric and with args like `some_func(x, foo=1)`: OK and not tested.
> 
> Emm, is it a typo here? not tested or tested?

Ah, typo error, it is tested. Changed. 

>
> I agree, you may ignore the 4th case right now. Please write a 'TODO' in the code and a note in the documentation that warns users to avoid this case.

@warshallrho Please write 'TODO' in the code and more importantly, add __Note__ for `Lambda` layer to warn users to avoid this case. When you finish, please let me and @JingqingZ know so we can approve and merge this PR.

Comment (warshallrho) [reply tests/models/test_model_save_graph.py]: got it

File: tensorlayer/layers/lambda_layers.py [hunk 10]
```diff
@@ -41,17 +41,35 @@ class Lambda(Layer):
 
     Examples
     ---------
-    Non-parametric case
+    Non-parametric with args case
```
Comment (JingqingZ) [inline tensorlayer/layers/lambda_layers.py]: This should be `without args`?

File: tensorlayer/layers/lambda_layers.py [hunk 11]
```diff
@@ -41,17 +41,35 @@ class Lambda(Layer):
...
+    In this case, you can use Model.save() / Model.load() to save / load the whole model architecture and weights(optional).
 
     >>> x = tl.layers.Input([8, 3], name='input')
     >>> y = tl.layers.Lambda(lambda x: 2*x, name='lambda')(x)
 
+    Non-parametric and non-args case
```
Comment (JingqingZ) [inline tensorlayer/layers/lambda_layers.py]: This should be `with args`?

File: tensorlayer/layers/lambda_layers.py [hunk 12]
```diff
@@ -41,17 +41,35 @@ class Lambda(Layer):
 
     Examples
     ---------
-    Non-parametric case
+    Non-parametric with args case
+    In this case, you can use Model.save() / Model.load() to save / load the whole model architecture and weights(optional).
```
Comment (JingqingZ) [inline tensorlayer/layers/lambda_layers.py]: The passive expression is prefered in formal language:

`This case is supported in the Model.save() / Model.load() to save / load the whole model architecture and weights(optional).`

File: tensorlayer/layers/lambda_layers.py [hunk 13]
```diff
@@ -41,17 +41,35 @@ class Lambda(Layer):
...
 
     >>> x = tl.layers.Input([8, 3], name='input')
     >>> y = tl.layers.Lambda(lambda x: 2*x, name='lambda')(x)
 
+    Non-parametric and non-args case
+    In this case, you can use Model.save() / Model.load() to save / load the whole model architecture and weights(optional).
```
Comment (JingqingZ) [inline tensorlayer/layers/lambda_layers.py]: Should be changed to passive expression as mentioned above.

File: tensorlayer/layers/lambda_layers.py [hunk 14]
```diff
@@ -41,17 +41,35 @@ class Lambda(Layer):
...
+
     >>> def customize_func(x, foo=42): # x is the inputs, foo is an argument
     >>>     return foo * x
     >>> x = tl.layers.Input([8, 3], name='input')
     >>> lambdalayer = tl.layers.Lambda(customize_func, fn_args={'foo': 2}, name='lambda')(x)
 
+
+    Parametric self defined function with outside variables
```
Comment (JingqingZ) [inline tensorlayer/layers/lambda_layers.py]: Does it have to be a `parametric` function or any function with outside variables?

File: tensorlayer/layers/lambda_layers.py [hunk 15]
```diff
@@ -41,17 +41,35 @@ class Lambda(Layer):
...
+    In this case, we hasn't supported using Model.save() / Model.load() to save / load the whole model architecture and weights(optional) yet.
+    Please avoid using Model.save() / Model.load(). Instead, you can use Model.save_weights() / Model.load_weights() to save / load model weights.
+    Note: In this case, fn_weights should be a list, and then the trainable weights in this LambdaLayer can be added into the weights of the whole model.
+
```
Comment (JingqingZ) [inline tensorlayer/layers/lambda_layers.py]: This case has not been supported in Model.save() / Model.load() yet.
Please avoid using Model.save() / Model.load() to save / load models that contain such Lambda layer. Instead, you may use Model.save_weights() / Model.load_weights() to save / load model weights.
Note: In this case, fn_weights should be a list, and then the trainable weights in this Lambda layer can be added into the weights of the whole model.

File: tensorlayer/layers/lambda_layers.py [hunk 16]
```diff
@@ -41,17 +41,35 @@ class Lambda(Layer):
...
+    >>> vara = [tf.Variable(1.0)]
+    >>> def func(x):
+    >>>     return x + vara
+    >>> x = tl.layers.Input([8, 3], name='input')
+    >>> y = tl.layers.Lambda(func, fn_weights=a, name='lambda')(x)
+
+
     Parametric case, merge other wrappers into TensorLayer
+    In this case, you can use Model.save() / Model.load() to save / load the whole model architecture and weights(optional).
```
Comment (JingqingZ) [inline tensorlayer/layers/lambda_layers.py]: Should be changed to passive expression as mentioned above.

File: tensorlayer/models/core.py [hunk 17]
```diff
@@ -710,6 +710,7 @@ def save(self, filepath, save_weights=True):
         >>> net.save('./model.h5', save_weights=True)
         >>> new_net = Model.load('./model.h5', load_weights=True)
         """
+        # TODO: support saving LambdaLayer that includes parametric self defined function with outside variables
```
Comment (JingqingZ) [inline tensorlayer/models/core.py]: In the documentation of Model.save() and Model.load(), we should ask users to check the documentation of Lambda layer if the model contains Lambda layer. For example

```python
def save(self, filepath, save_weights=True):
        """
        Save model into a given file.
        This function can save both the architecture of neural networks and weights (optional).
        WARNING: If the model contains Lambda layer, please check the documentation of Lambda layer and find out the cases that have / have not been supported by Model.save().

        ...
        """
```

Please also include the warning information in `Model.load()`.

Comment (warshallrho) [reply tensorlayer/models/core.py]: I have got the above mentioned comments.  Thanks a lot. I will fix them and test ElementwiseLambda soon.
